File: 백준/Gold/14567. 선수과목 （Prerequisite）/선수과목 （Prerequisite）.py
import sys
input = sys.stdin.readline
from collections import deque


def solution():
    N, M = map(int, input().split())

    answer = [1]*N
    parent = [[] for _ in range(N)]

    for _ in range(M):
        A, B = map(int, input().split())
        parent[B-1].append(A-1)



    # 선수 과목은 i보다 인덱스가 작을 수 밖에 없음
    for i in range(N):
        for p in parent[i]:
            answer[i] = max(answer[p]+1, answer[i])

    


    for k in answer:
        print(k, end=' ')

if __name__ == "__main__":
    solution()
    

    

File: 백준/Gold/14567. 선수과목 （Prerequisite）/선수과목 （Prerequisite）.cc
#include <iostream>
#include <string>
#include <vector>
#include <numeric>
using namespace std;

void solution() {
    int N, M, A, B;
    cin >> N >> M;

    vector<int> answer(N, 1);
    vector<vector<int>> parent(N);


    for (int i = 0; i < M; i++) {
        cin >> A >> B;
        parent[B - 1].push_back(A - 1);
    }

    for (int i = 0; i < N; i++) {
        for (int p : parent[i]) {
            answer[i] = max(answer[p] + 1, answer[i]);
        }
    }

    
    for (int k : answer) {
        cout << k << ' ';
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    solution();

    return 0;
}